### ISSUE TITLE
Unit-typed properties compile to void get methods

### ISSUE THREAD

**Issue by kqr** [dotnet/fsharp#9298]
I'm not 100% confident where this bug is, or if I'm misunderstanding something, but there's something odd going on, somewhere, at least.

**Repro steps**

Create a class with a unit-typed property, like so:

```fsharp
type MyClass() =
    member this.MyProperty =
        ()
```

**Expected behavior**

I sorta kinda expect F# to follow C#s lead here, and reject this construction at compile-time, because a unit-typed property makes only a sliver of sense.

**Actual behavior**

The compiler gives me some IL where the following is the interesting bit:

```
        .method public hidebysig specialname 
            instance void get_MyProperty () cil managed 
        {
            // Method begins at RVA 0x205c
            // Code size 1 (0x1)
            .maxstack 8

            IL_0000: ret
        } // end of method MyClass::get_MyProperty

        // Properties
        .property instance class [FSharp.Core]Microsoft.FSharp.Core.Unit MyProperty()
        {
            .get instance void _/MyClass::get_MyProperty()
        }
```

This does not play well with some tools, like the VS debugger which crashes when trying to inspect an object with this property. 

I suspect the offending part might be how a property of type `instance class Unit` has a getter returning `instance void`.

**Known workarounds**

One possible "workaround" is to turn the property into a regular method. This will be a void method that all the tooling already understands.

**Comment by KevinRansom** [comment]
@kqr, this construct crashes VS and so we need to do something that stops that.  Changed the title to a much scarier one so that I remember that we should do something about this.

**Comment by KevinRansom** [comment]
I looked at how C# would treat this Property:
It produces this error:
![image](https://user-images.githubusercontent.com/5175830/83291677-93cf5f80-a19d-11ea-81e5-b6ca8d523523.png)

**Comment by KevinRansom** [comment]
So ... 
````
module public Say
    let z = ()
````

Is emitted as an FSharp.Unit
````
  .method public specialname static class [FSharp.Core]Microsoft.FSharp.Core.Unit 
          get_z() cil managed
  {
    .custom instance void [netstandard]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = ( 01 00 00 00 ) 
    .custom instance void [netstandard]System.Diagnostics.DebuggerNonUserCodeAttribute::.ctor() = ( 01 00 00 00 ) 
    // Code size       2 (0x2)
    .maxstack  8
    IL_0000:  ldnull
    IL_0001:  ret
  } // end of method Say::get_z
````
So it seems as if fixing the codegen for the property getter to match let public z = () I.e returning a null typed as Unit would be good.

I don't think that would be a breaking change, C# code could not have assigned the value of the property, and would have to have invoked the get_ function, to achieve whatever side-effects they were dependent upon.  Although now a unit is returned that value is just ignored by the C# compiler.  I believe the F# compiler will treat a property returning unit correctly.  The debugger should just display the value null for the property in watch windows, which is not terrible.

@dsyme / @TIHan , do you know why the unit property codegen implements a unit returning property getter as void returning?

**Comment by abelbraaksma** [comment]
Mainly out of curiosity, I tried this back to FSC 3.0. All versions behave the same w.r.t. a unit-property. I didn't check on each version whether debugging would crash it though.

I doubt fixing this will have a big impact (I mean, aside from not crashing the IDE anymore), but if, for whatever reason, such properties are currently discovered with Reflection, getting the property would be discovered as having `void`, and getting the underlying method `get_MyProperty` would be discovered as having `Unit`. In other words, after fixing this, code that depends on this (why would it?) will fail.

**Comment by abelbraaksma** [comment]
> and would have to have invoked the get_ function, to achieve whatever side-effects they were dependent upon

Just noted that C# refuses to assign the result of the gettor, because it considers it `void` when using `get_XXX` directly. I managed to compile it only if I don't assign it (though it stays red in the IDE). If you just call the gettor without assigning it, it's fine. 

Which basically suggests: anyone seriously using `unit` as a result of a property must have run into this, and then stayed away from it. Imo: fixing it to both be `unit` will make it possible to assign the result from C# to a variable.

**Comment by abelbraaksma** [comment]
Behavior of C# seems to have changed now. For a property with a single `ret` in the IL:

* C# doesn't crash when stopping at a breakpoint
* If it's the only property, it won't show it when hovering while debugging
* If there are more properties, it shows the other properties, but not the `void` one in the context-sensitive help
* You cannot access the property normally, only with the `get_` prefix, like `x.get_MyVoidProperty()`
* When you try to assign the property to a `var` or `_` in C#, it says: 
  > C8209: A value of type 'void' may not be assigned
* You can assign the property as an `Action` just fine

Also, using it in F# goes without problems, and debugging (in Debug mode, otherwise it's inlined) goes fine.

Note that the body of this property is still emitted the same way as described by the OP. It's just that the C#-compiler/debugger behaves differently.

TLDR: there are still some oddities, but it doesn't crash VS anymore while debugging. Not sure about changing F# to output different IL. It still makes sense to be orthoganal here and do the same as in other F# functions and emit what @KevinRansom describes above:

```
    IL_0000:  ldnull
    IL_0001:  ret
```
